Implementation for /most/ of SE-0192 (frozen and non-frozen enums)

jrose-apple opened the pull request:
An alternative to #11961 that only makes a frozen / non-frozen distinction for imported C enums and for Swift enums in the stdlib/overlays. (Really, it's "libraries compiled with `-enable-resilience`", and exactly what that means has not been discussed publicly and is not part of SE-0192.)

This does not include the implementation of `unknown case`, or whatever we end up naming it. That's a big enough change on its own that it deserves its own PR, and this part can land first.

The diagnostics part of this is behind a frontend flag, `-enable-nonfrozen-enum-exhaustivity-diagnostics`.

jrose-apple committed:
Add Parse + Sema support for '@_frozen'

(currently spelled with an underscore to indicate its WIP state)

Later commits will handle imported enums correctly and implement the
checks for switch cases.

jrose-apple committed:
Diagnose uncovered switches on non-frozen enums

Warn in Swift 4 mode and error in Swift 5 mode when switching on a
non-frozen enum without providing a default case.

Note that this is a preliminary implementation, in order to test the
rest of the feature.

jrose-apple committed:
Switch exhaustiveness checking: SpaceKind is an enum, not option set.

No functionality change.

jrose-apple committed:
Treat exhaustive enums as "non-resilient", and add a few more tests

"Formally non-resilient" in this new world means "the enum has a fixed
representation", which implies a fixed layout algorithm. We're not
there yet, but non-exhaustive enums should be able to be fixed-layout
as well by picking a general representation that won't need to grow.
Specifically, that's enums with raw types, and possibly also indirect
enums as well.

(It's likely the '_fixed_layout' /attribute/ on enums will go away,
but the concept of a fixed-layout enum is still useful.)

jrose-apple committed:
[ClangImporter] Prefer later enum_extensibility over earlier ones

Since NS_ENUM has an enum_extensibility(open) in it already, we want
to allow people to undo that by sticking enum_extensibility(closed) on
the end of their enum.

jrose-apple committed:
[ClangImporter] Translate Clang's enum_extensibility to @_frozen

Still to do: test and fix up the use of multiple enum_extensibility
annotations, possibly with API notes. This is important because the
definition of NS/CF_ENUM /includes/ enum_extensibility(open) as of
Xcode 9.0; there should be a convenient out people can use to declare
exhaustive enums in C that's backwards-compatible.

jrose-apple committed:
Put non-frozen enum exhaustivity diagnostics behind a frontend flag

...spelled '-enable-nonfrozen-enum-exhaustivity-diagnostics'. This
is for staging purposes.

jrose-apple committed:
[PrintAsObjC] Use enum_extensibility to represent @_frozen

Previously (a03c40cb2c) we assumed all Swift enums were non-frozen in
ObjC, a weird choice in retrospect. Now that we actually distinguish
frozen and non-frozen enums in Swift, we can use the
'enum_extensibility' attribute to mark them as open or closed in ObjC.

Note that this only matters for Swift libraries compiled with
-enable-resilience, i.e. those that might get a new implementation at
runtime. Everyone else is now declaring a "closed" enum, matching the
behavior in Swift.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple committed:
Ban @_fixed_layout on enums in favor of @_frozen

In theory there could be a "fixed-layout" enum that's not exhaustive
but promises not to add any more cases with payloads, but we don't
need that distinction today.

(Note that @objc enums are still "fixed-layout" in the actual sense of
"having a compile-time known layout". There's just no special way to
spell that.)

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please clean test macOS

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
- [Compiler performance](https://ci.swift.org/view/Pull%20Request/job/swift-PR-compiler-performance-macOS/111/) (still running)
- [Source compat suite](https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite/827/) (passed)

jrose-apple commented:
preset=buildbot_all_platforms,tools=RA,stdlib=RA,resilience
@swift-ci Please test with preset

jrose-apple commented:
preset=buildbot_all_platforms,tools=RA,stdlib=RA,resilience
@swift-ci Please test macOS with preset

jrose-apple commented:
preset=buildbot_all_platforms,tools=RA,stdlib=RA,resilience
@swift-ci Please test with preset

jrose-apple commented:
- [Preset build](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-preset/6/) (though on the wrong commit, so I'm not sure what's going to happen)

jrose-apple commented:
Hm. Well, at least…

@swift-ci Please smoke test compiler performance

jrose-apple commented:
@DougGregor or @rjmccall, can you review the ClangImporter parts? @slavapestov, can you review the resilience parts?

(I'm not worrying about the switch exhaustivity part because `unknown case` rips out most of what I changed anyway. I'll ask Robert to review when we're ready for that.)

jrose-apple commented:
Rebased.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

DougGregor commented on stdlib/public/core/Codable.swift.gyb:
```diff
@@ -1101,7 +1101,7 @@ public struct CodingUserInfoKey : RawRepresentable, Equatable, Hashable {
 
 /// An error that occurs during the encoding of a value.
-@_fixed_layout // FIXME(sil-serialize-all)
+@_frozen // FIXME(sil-serialize-all)
```
This is a surprise to me; why should encoding and decoding errors be frozen? I'd expect this to grow over time, like most error types.

jrose-apple replied:
They'll definitely get un-frozen. This commit did not remove any `frozen`s that were currently there.

slavapestov replied:
... it did in stdlib/public/Platform/POSIXError.swift

jrose-apple replied:
…oops. Okay, I'll be consistent.

DougGregor commented on stdlib/public/core/DebuggerSupport.swift:
```diff
@@ -10,9 +10,9 @@
-  @_fixed_layout // FIXME(sil-serialize-all)
+  @_frozen // FIXME(sil-serialize-all)
   @_versioned // FIXME(sil-serialize-all)
   internal enum CollectionStatus {
```
Is there any point to a frozen internal enum?

jrose-apple replied:
It's versioned, so yes.

DougGregor commented on test/ClangImporter/Inputs/custom-modules/EnumExhaustivity.apinotes:
```diff
@@ -0,0 +1,4 @@
+Name: EnumExhaustivity
+Tags:
+- Name: RegularEnumTurnedExhaustiveThenBackViaAPINotes
+  EnumExtensibility: open
```
I'd never seen this cute "no newline at end of file" GitHub graphic before

DougGregor commented on include/swift/Basic/LangOptions.h:
```diff
@@ -239,6 +242,10 @@ namespace swift {
 
+    /// Diagnose switches over non-frozen enums that do not have catch-all
+    /// cases.
+    bool EnableNonFrozenEnumExhaustivityDiagnostics = false;
```
This is for staging while we settle on the "unknown case" spelling, right? It'll go away then?

jrose-apple replied:
It's mainly so I can land this part before SE-0192's review period ends, but yes.

DougGregor reviewed (approved):
This looks good. The space engine part is appropriately mind-boggling, but I've convinced myself it looks good. Nice work!

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -3088,6 +3090,53 @@ bool EnumDecl::hasOnlyCasesWithoutAssociatedValues() const {
+
+  // All other checks are use-site specific; with no further information, the
+  // enum must be treated non-exhaustively.
+  if (!useDC)
```
Do we ever need to check exhaustiveness in SILGen or anything else that runs after Sema? Because we won't have a useDC anymore

jrose-apple replied:
I think we'll have a DC up through SILGen, but you do have me worried about optimizations. Unfortunately we don't actually the right API for that ("AvailabilityContext"); fortunately, it'll err on the side of "non-exhaustive". (That is, passing `nullptr` or the containing module or the containing source file might answer "non-exhaustive" when it could be "exhaustive", but never the other way around.)

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2102,6 +2105,27 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
+void AttributeChecker::visitFrozenAttr(FrozenAttr *attr) {
+  switch (D->getModuleContext()->getResilienceStrategy()) {
+  case ResilienceStrategy::Default:
+    // FIXME: Don't warn about the use of this attribute in the standard
```
You might be able to get away with removing this now that resilience is on

jrose-apple replied:
Ah, yes, I forgot about this!

slavapestov commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -121,25 +127,25 @@ namespace {
       static const size_t MAX_SPACE_SIZE = 128;
 
-      size_t computeSize(TypeChecker &TC,
+      size_t computeSize(TypeChecker &TC, const DeclContext *DC,
```
It's awkward to plumb the DC (and TC) through everything here -- can't we make them instance variables?

jrose-apple replied:
Sure, that makes sense. I'll save it for the next half, though, where I implement `unknown:`.

slavapestov reviewed (commented):
Please add some code generation (SILGen) tests for -enable-testing. I imagine we'll want to trap if the original library evolves resiliently, right?

Also, library evolution tests for non-frozen @objc enums to ensure the unknown case is handled properly.

jrose-apple committed:
[stdlib] Remove '@_frozen' from enums that shouldn't be frozen

Error codes, FloatingPointRoundingRule, Mirror.AncestorRepresentation,
Mirror.DisplayStyle, and PlaygroundQuickLook.

jrose-apple committed:
[stdlib] Add '@_frozen' to overlay enums that need it

That's just ARCamera.TrackingState and DispatchTimeoutResult.
Verified with the framework owners at Apple.

jrose-apple commented:
We do have the "evolution" tests for non-frozen enums by testing private cases, but I'll add the ones for `-enable-testing`!

slavapestov commented:
@jrose-apple The case I'm thinking of is when you switch over an @objc non-frozen enum, we want to make sure we don't hit UB if you pass in an unknown case

jrose-apple commented:
Sorry, I forgot that one already landed. :-) See #14724.

jrose-apple committed:
Add tests for -enable-testing silencing exhaustivity diagnostics

jrose-apple commented:
Addressed feedback.

@swift-ci Please smoke test

jrose-apple commented:
Merging even though the proposal is still in review in order to get some living-on (and so that I don't have to keep rebasing it). Note that the diagnostics are still off and `unknown` is not yet implemented. (That's #14382, which I'm currently working on rebasing.)

patrickbdev commented:
Is there a way to suppress the warning this produces if you don't have a `@unknown default` in your switch?

jrose-apple commented:
I'm not working on Swift anymore, but people have talked about this for "I know this library is built with library evolution turned on, but I'm embedding it in my app, so it can't possibly change out from under me". It's not clear that such an annotation would really be any better than just writing the `@unknown default` though.

jrose-apple commented:
I do suggest taking this discussion to the [forums](https://forums.swift.org) rather than commenting on a two-month-old PR, if you haven't already. (As you noticed, it took a month for me to even see it!)
